src/query/api/v1/handler/prometheus/remote/write.go

```diff
@@ -555,6 +570,72 @@ func (h *PromWriteHandler) forward(
 	return nil
 }
 
+func (h *PromWriteHandler) maybeLogSeriesWithLongestLabels(logger *zap.Logger, timeSeries []prompb.TimeSeries) {
+	if len(timeSeries) == 0 {
+		return
+	}
+
+	if atomic.AddUint32(&h.numLiteralIsTooLong, 1) > maxLiteralIsTooLongLogCount {
+		return
+	}
+
+	var (
+		longestLabelLength = 0
+		longestLabelIdx    = 0
+	)
+	for i, ts := range timeSeries {
+		for _, l := range ts.Labels {
+			if len(l.Name) > longestLabelLength {
+				longestLabelLength = len(l.Name)
+				longestLabelIdx = i
+			}
+			if len(l.Value) > longestLabelLength {
+				longestLabelLength = len(l.Value)
+				longestLabelIdx = i
+			}
+		}
+	}
+
+	const totalLengthThreshold = 1000
+
+	labelSummary := labelsToSummaryString(timeSeries[longestLabelIdx].Labels, totalLengthThreshold)
+	logger.Warn("series with too long literal", zap.String("labelSummary", labelSummary))
+}
+
+func labelsToSummaryString(labels []prompb.Label, lengthThreshold int) string {
+	builder := strings.Builder{}
+	builder.WriteRune('{')
+	for i, l := range labels {
+		if !appendLiteral(&builder, l.Name, lengthThreshold) {
+			break
+		}
+		builder.WriteRune('=')
+		if !appendLiteral(&builder, l.Value, lengthThreshold) {
+			break
+		}
+		if i+1 < len(labels) {
+			builder.WriteRune(',')
+		}
+	}
+	if builder.Len() < lengthThreshold {
+		builder.WriteRune('}')
+	} else {
+		builder.WriteString("...")
+	}
+	return builder.String()
+}
+
+func appendLiteral(builder *strings.Builder, literal []byte, lengthThreshold int) bool {
+	if builder.Len()+len(literal) <= lengthThreshold {
+		builder.Write(literal)
+		return true
+	}
+
+	n := lengthThreshold - builder.Len()
+	builder.Write(literal[:n])
+	return false
+}
```

Comment:
Instead of all this, can we simply log a prefix of the first offending value?

Reply:
https://github.com/m3db/m3/pull/3801/commits/47bdec88d34ccd484db3259e2fd4e8559ac94715

Kept the code that selects the longest literal, since in theory the literal length limit is configurable, so it's not exactly clear which would be the first offending value.

https://github.com/m3db/m3/blob/066d6e374b21f702c752cdbbb2807a49f7f6abea/src/x/serialize/encoder.go#L185

Reply:
As discussed offline, added `MaxTagLiteralLength()` method to `TagOptions` and moved long literal check to where the request is parsed.

https://github.com/m3db/m3/pull/3801/commits/53f179427919c220d17382fd7b30864c0e5016d4

However, this also prompted adding additional code in `src/query/models/tags.go`, since it also does tag validation according to values in `TagOptions`.